core/src/main/clojure/xtdb/expression/temporal.clj

```diff
@@ -460,6 +460,107 @@
+  (let [period (.getPeriod x)
+        sign (if (.isNegative period) "-" "")
+        abs-total-months (abs (.toTotalMonths period))
+        year (quot abs-total-months 12)
+        month (mod abs-total-months 12)
+        interval-string (cond-> ""
+                          (not= year 0) (str year (when (not= month 0) "-"))
+                          (not= month 0) (str month))
+        interval-qualifier (cond-> ""
+                             (not= year 0) (str "YEAR" (when (not= month 0) " TO "))
+                             (not= month 0) (str "MONTH"))]
+    (format "INTERVAL '%s%s' %s" sign interval-string interval-qualifier)))
+
+(defn- get-present-fields [days hours minutes seconds nanos]
+  (cond-> []
+    (not= days 0) (conj {:field "DAY" :rank 1})
+    (not= hours 0) (conj {:field "HOUR" :rank 2})
+    (not= minutes 0) (conj {:field "MINUTE" :rank 3})
+    (not= seconds nanos 0) (conj {:field "SECOND" :rank 4})))
+
+(defn- format-end-field [start-field end-field nanos?]
+  (cond
+    (and (= start-field end-field) nanos?) "SECOND(1,9)"
+    nanos? "SECOND(9)"
+    :else end-field))
+
+(defn- ->interval-str [days hours minutes seconds nanos start-rank end-rank]
+  (cond-> ""
+    (= start-rank 1) (str (abs days) (when (> end-rank 1) " "))
+    (and (<= start-rank 2) (> end-rank 1)) (str (abs hours) (when (> end-rank 2) ":"))
+    (and (<= start-rank 3) (> end-rank 2)) (str (abs minutes) (when (> end-rank 3) ":"))
+    (and (<= start-rank 4) (> end-rank 3)) (str (abs seconds) (when (> nanos 0) (format ".%09d" (abs nanos))))))
+
+(defn mdn-interval->string [^PeriodDuration x]
+  (let [period (.getPeriod x)]
+    (if (> (.toTotalMonths period) 0)
+      (throw (err/runtime-err :xtdb.expression/cannot-cast-mdn-interval-with-months
+                              {::err/message "Cannot cast month-day-nano intervals when month component is non-zero."}))
+      (let [duration (.getDuration x)
+            days (.getDays period)
+            sign (if (or (neg? days) (.isNegative duration)) "-" "")
+            abs-duration (.abs duration)
+            hours (.toHoursPart abs-duration)
+            minutes (.toMinutesPart abs-duration)
+            seconds (.toSecondsPart abs-duration)
+            nanos (.toNanosPart abs-duration)
+            present-fields (get-present-fields days hours minutes seconds nanos)
+            {start-field :field start-rank :rank} (first present-fields)
+            {end-field :field end-rank :rank} (last present-fields)
+            formatted-end-field (format-end-field start-field end-field (> nanos 0))
+            interval-str (->interval-str days hours minutes seconds nanos start-rank end-rank)
+            interval-qualifier (if (= start-field end-field) formatted-end-field (format "%s TO %s" start-field formatted-end-field))]
+        (format "INTERVAL '%s%s' %s" sign interval-str interval-qualifier)))))
+
+(defmethod expr/codegen-cast [:interval :utf8] [{[_ iunit] :source-type}]
```

Comment:
given that the new proposal is to return ISO8601 here, I might be tempted to migrate this pre-PR to cut down on the churn? (not a strongly-held opinion)